lib/ui/compositing/scene_builder.cc

```diff
@@ -97,41 +154,63 @@ void SceneBuilder::pushShaderMask(Shader* shader,
                                   double maskRectTop,
                                   double maskRectBottom,
                                   int blendMode) {
-  layer_builder_->PushShaderMask(
-      shader->shader(),
-      SkRect::MakeLTRB(maskRectLeft, maskRectTop, maskRectRight,
-                       maskRectBottom),
-      static_cast<SkBlendMode>(blendMode));
+  SkRect rect = SkRect::MakeLTRB(maskRectLeft, maskRectTop, maskRectRight,
+                                 maskRectBottom);
+  auto layer = std::make_unique<flow::ShaderMaskLayer>();
+  layer->set_shader(shader->shader());
+  layer->set_mask_rect(rect);
+  layer->set_blend_mode(static_cast<SkBlendMode>(blendMode));
+  PushLayer(std::move(layer), cull_rects_.top());
 }
 
 void SceneBuilder::pushPhysicalShape(const CanvasPath* path,
                                      double elevation,
                                      int color,
                                      int shadow_color,
-                                     int clip_behavior) {
-  layer_builder_->PushPhysicalShape(
-      path->path(),                 //
-      elevation,                    //
-      static_cast<SkColor>(color),  //
-      static_cast<SkColor>(shadow_color),
-      UIDartState::Current()->window()->viewport_metrics().device_pixel_ratio,
-      static_cast<flow::Clip>(clip_behavior));
+                                     int clipBehavior) {
+  const SkPath& sk_path = path->path();
+  flow::Clip clip_behavior = static_cast<flow::Clip>(clipBehavior);
+  SkRect cullRect;
+  if (!cullRect.intersect(sk_path.getBounds(), cull_rects_.top())) {
+    cullRect = SkRect::MakeEmpty();
+  }
+  auto layer = std::make_unique<flow::PhysicalShapeLayer>(clip_behavior);
+  layer->set_path(sk_path);
+  layer->set_elevation(elevation);
+  layer->set_color(static_cast<SkColor>(color));
+  layer->set_shadow_color(static_cast<SkColor>(shadow_color));
+  layer->set_device_pixel_ratio(
+      UIDartState::Current()->window()->viewport_metrics().device_pixel_ratio);
+  PushLayer(std::move(layer), cullRect);
 }
 
 void SceneBuilder::pop() {
-  layer_builder_->Pop();
+  if (!current_layer_) {
+    return;
+  }
+  cull_rects_.pop();
+  current_layer_ = current_layer_->parent();
 }
 
 void SceneBuilder::addPicture(double dx,
                               double dy,
                               Picture* picture,
                               int hints) {
-  layer_builder_->PushPicture(
-      SkPoint::Make(dx, dy),                             //
-      UIDartState::CreateGPUObject(picture->picture()),  //
-      !!(hints & 1),                                     // picture is complex
-      !!(hints & 2)                                      // picture will change
-  );
+  if (!current_layer_) {
+    return;
+  }
+  SkPoint offset = SkPoint::Make(dx, dy);
+  SkRect pictureRect = picture->picture()->cullRect();
```

Comment:
I think this two lines of code is useless since line 205-207 was deleted in some commit.